File: app/src/main/java/xyz/muggr/newsly/Managers/ArticleQueueManager.java
package xyz.muggr.newsly.Managers;

import android.database.sqlite.SQLiteDatabase;
import android.os.AsyncTask;
import android.util.Log;

import xyz.muggr.newsly.Articles.ArticleList;
import xyz.muggr.newsly.BuildConfig;
import xyz.muggr.newsly.NewslyActivity;

public class ArticleQueueManager {

    private ArticleList articleQueue;

    private NewslyActivity activity;
    private final SQLiteDatabase database;

    //region Constructors
    //=======================================================================================

    public ArticleQueueManager(NewslyActivity activity) {
        this.activity = activity;
        database = DatabaseManager.getInstance(activity).getWritableDatabase();
        articleQueue = new ArticleList();
    }

    //=======================================================================================
    //endregion

    //region Article queue modifiers
    //=======================================================================================

    public void load(final ArticleQueueListener listener) {

        if (articleQueue.size() < 10)
            new AsyncTask<Void, Void, ArticleList>() {
                @Override
                protected ArticleList doInBackground(Void... params) {
                    try {
                        return ApiManager.getArticles();
                    } catch (Exception e) {
                        if (BuildConfig.DEBUG)
                            e.printStackTrace();
                        return null;
                    }
                }

                @Override
                protected void onPostExecute(ArticleList articleList) {
                    super.onPostExecute(articleList);
                    if (articleList != null) {
                        articleQueue.addAll(articleList);
                    }
                    listener.onArticleQueueLoaded(articleQueue);

                    if (BuildConfig.DEBUG)
                        for (int i = 0; i < articleQueue.size(); i++) {
                            Log.d(
                                    String.format("Queue %02d", i),
                                    articleQueue.get(i).getHeadline()
                            );
                        }
                }

            }.execute();
        else
            listener.onArticleQueueLoaded(articleQueue);

    }

    public void dismiss(int state) {
        if (!articleQueue.isEmpty()) {
            articleQueue.remove(0);
        }
    }

    //=======================================================================================
    //endregion

    //region Interfaces
    //=======================================================================================

    public interface ArticleQueueListener {
        void onArticleQueueLoaded(ArticleList articleQueue);
    }

    //=======================================================================================
    //endregion

}

File: app/src/main/java/xyz/muggr/newsly/Managers/ArticleQueueManager.kt
package xyz.muggr.newsly.Managers

import android.database.sqlite.SQLiteDatabase
import android.os.AsyncTask
import xyz.muggr.newsly.Articles.ArticleList
import xyz.muggr.newsly.BuildConfig
import xyz.muggr.newsly.NewslyActivity

class ArticleQueueManager(private val activity: NewslyActivity) {

    private var articleQueue: ArticleList
    private val database: SQLiteDatabase

    init {
        database = DatabaseManager.getInstance(activity).writableDatabase
        articleQueue = ArticleList()
    }

    //region Article queue modifiers
    //=======================================================================================

    fun load(listener: ArticleQueueListener) {
        GetArticleListTask(articleQueue, listener).execute()
    }

    fun dismiss(state: Int) {
        if (!articleQueue.isEmpty()) {
            articleQueue.removeAt(0)
        }
    }

    //=======================================================================================
    //endregion

    //region Asynctasks
    //=======================================================================================

    class GetArticleListTask(private val articleQueue: ArticleList, private val listener: ArticleQueueListener) : AsyncTask<Void, Void, ArticleList>() {

        override fun doInBackground(vararg params: Void): ArticleList? {
            try {
                return ApiManager.getCards()
            } catch (e: Exception) {
                if (BuildConfig.DEBUG)
                    e.printStackTrace()
                return null
            }

        }

        override fun onPostExecute(articleList: ArticleList?) {
            super.onPostExecute(articleList)
            if (articleList != null) {
                articleQueue.addAll(articleList)
            }
            listener.onArticleQueueLoaded(articleQueue)
        }

    }

    //=======================================================================================
    //endregion

    //region Interfaces
    //=======================================================================================

    interface ArticleQueueListener {
        fun onArticleQueueLoaded(articleQueue: ArticleList)
    }

    //=======================================================================================
    //endregion

}
